Add General parshift demo

dwierichs opened the pull request:
This demo implements the univariate reconstruction ideas from [this paper](https://arxiv.org/abs/2107.12390), for the case of equidistant frequencies in cost functions (the main text of the paper focuses on that case as well).
Furthermore, it showcases the generalized parameter-shift rules derived there, using equidistant shifts.
Finally, Rotosolve is implemented in a minimal fashion and applied to a toy circuit with multiple frequencies per parameter.

dwierichs commented:
@mariaschuld Would you be interested in reviewing this next week?

josh146 commented on demonstrations/tutorial_general_parshift.py:
```diff
@@ -0,0 +1,833 @@
+
+*Author: David Wierichs (Xanadu resident). Posted: ?? August 2021.*
+
+In this demo we will learn how to reconstruct univariate quantum functions :math:`E` that are
+generated by parametrized (layers of) gates and how to derive
+:doc:`parameter-shift rules </glossary/parameter_shift>` to compute the derivatives of any order
+of such functions.
+All we will need for this is the insight that these functions of interest that only depend on
+a single variable :math:`x` are Fourier series in this variable and a bit of knowledge about
+Fourier transforms. In the case of general shifts for the shift rules, linear algebra actually
+is all we use for the reconstruction.
+We briefly recap the derivation of the functional form of :math:`E(x)` further below, the full
+version can be found together with details on the reconstruction idea, derivations of the
+parameter-shift rules, and considerations for multivariate functions in [#GenPar]_.
+The core idea was also presented in [#CalcPQC]_ but with fewer conclusions drawn from it and
+different shift rules.
+We will follow [#GenPar]_ regarding the contents and notation but also point out two more
+preprints discussing general parameter-shift rules: an algebraic approach [#AlgeShift]_ and
+one focusing on special gates and spectral decompositions [#GenDiffRules]_.
```
A couple of general thoughts:

- I like how the intro straight away says what the reader will learn, this is great!

- I'm a big fan of having direct hyperlinks to papers (alongside a list of references). This way it is easy to scan the intro, see a paper link, and open it in a new tab. E.g., see the [intro to the GBS demo](https://pennylane.ai/qml/demos/tutorial_gbs.html), or the [doubly stochastic demo](https://pennylane.ai/qml/demos/tutorial_doubly_stochastic.html).

- I think it will also be worthwhile to simplify the intro significantly, and remove all technicalities. Simply rephrase it in simply terms; what will we learn, why is this interesting, and what can we do with this?

dwierichs replied:
Gave it a shot :)
In particular, the reference subsection is a bit larger now, due to the inline paper refs, and I simplified the first part(s) and removed math object and technicalities.

josh146 replied:
Reads much better!

josh146 commented on demonstrations/tutorial_general_parshift.py:
```diff
@@ -0,0 +1,833 @@
+
+|
+
+.. figure:: ../demonstrations/general_parshift/thumbnail_genpar.png
+    :align: center
+    :width: 50%
+    :target: javascript:void(0)
```
This plot is so nice 🤩

dwierichs replied:
Haha thanks! "plot"... :D

josh146 commented on demonstrations/tutorial_general_parshift.py:
```diff
@@ -0,0 +1,833 @@
+cost_functions = []
+fig, axs = plt.subplots(1, len(Ns), figsize=(12, 2))
+X = np.linspace(*xlim, 60)
+green = "#209494"
+orange = "#ED7D31"
+red = "xkcd:brick red"
+blue = "xkcd:cerulean"
+pink = "xkcd:bright pink"
```
I wonder if it is worth moving this to a separate 'cell', so that it is easier to identify the important parts of the plotting code?

josh146 commented on demonstrations/tutorial_general_parshift.py:
```diff
@@ -0,0 +1,833 @@
+red = "xkcd:brick red"
+blue = "xkcd:cerulean"
+pink = "xkcd:bright pink"
+for ax, N in zip(axs, Ns):
+    # Generate the cost function for N qubits and plot it
+    dev, cost = device_and_cost(N, seed)
+    ax.plot(X, [cost(x) for x in X], color=green)
+    ax.set_title(f"{N} qubits")
+    ax.set_xlabel("$x$")
```
For pedagogical purposes, it might also be good to separate the E(x) reconstruction code from the plotting code 🤔

josh146 commented on demonstrations/tutorial_general_parshift.py:
```diff
@@ -0,0 +1,833 @@
+# Indeed we see that :math:`E(x)` is a periodic function whose complexity grows when increasing
+# the number of gates parametrized by :math:`x`.
+# We take a look at the frequencies that are supported by the functions using features from
+# PennyLane's ``fourier`` module.
```
```suggestion
# PennyLane's :mod:`~.pennylane.fourier` module.
```
This will ensure that it will turn into a hyperlink :)

josh146 commented on demonstrations/tutorial_general_parshift.py:
```diff
@@ -0,0 +1,833 @@
+    axs[1, i].text(Ns[i] + 2, axs[1, i].get_ylim()[0], f"Frequency", ha="center", va="top")
+    # Clean up y-axis labels
+    if i == 0:
+        _ = [axs[j, i].set_ylabel(lab) for j, lab in enumerate(["$a_\ell/2$", "$b_\ell/2$"])]
```
Maybe I missed it, but it doesn't look like `a_\ell` and `b_\ell` are defined anywhere?

dwierichs replied:
Oops!

josh146 commented on demonstrations/tutorial_general_parshift.py:
```diff
@@ -0,0 +1,833 @@
+#
+# .. math ::
+#
+#   U(x)=\prod_{a=1}^N R_Z^{(a)}(x) = \prod_{a=1}^N \exp\left(-i\frac{x}{2} Z_a\right)
```
I would probably at least copy this up earlier, to where the cost function is first defined :)

josh146 commented on demonstrations/tutorial_general_parshift.py:
```diff
@@ -0,0 +1,833 @@
+# As discussed in Sec. II A of the paper, the frequencies in the Fourier series will be the
+# *unique, positive* differences of these eigenvalues, which are:
+#
+# .. math ::
+#
+#   \{\Omega_\ell\} = \{1, 2,\dots, N\},
+#
+# as well as a zero frequency leading to a constant term in :math:`E(x)`.
+# This is exactly what we saw in the plots above.
+# In general we will call the number of these frequencies :math:`R`, so that we have :math:`R=N`
+# for the layer of ``RZ`` gates above.
```
It might be useful having a small codeblock here, showing the `\Omega_\ell` being computed for a particular generator?

dwierichs replied:
I'm wondering how simple this would be to showcase. I'll try something..

dwierichs replied:
Let me know whether the way I did it now is helpful :) @josh146

josh146 commented on demonstrations/tutorial_general_parshift.py:
```diff
@@ -0,0 +1,833 @@
+
+
+###############################################################################
+# We find the number of (positive) frequencies that appear in :math:`E(x)` to be the same as the
+# number of ``RZ`` gates we used in the circuit.
+# This is no coincidence:
```
It might be worth being more explicit here, since I can imagine there could be confusion (e.g., confusing 'positive' to refer to the bars above the horizontal, being confused over the two sets of plots, not realizing the constant term is not included, etc.)

dwierichs replied:
I feel like with the moving of the definition of $a_\ell$ and the earlier mention of why only positive frequencies are needed, this is taken care of?

josh146 commented on demonstrations/tutorial_general_parshift.py:
```diff
@@ -0,0 +1,833 @@
+# In general we will call the number of these frequencies :math:`R`, so that we have :math:`R=N`
+# for the layer of ``RZ`` gates above.
+#
+# The reason we can restrict ourselves to positive frequencies is that :math:`E(x)` is real-valued.
+# This is because :math:`B` is Hermitian, and it implies (anti-)symmetry for the real (imaginary)
+# Fourier coefficients, as one can also see in the spectral plots above.
```
I would move this much higher up!

josh146 commented on demonstrations/tutorial_general_parshift.py:
```diff
@@ -0,0 +1,833 @@
+# --------------------------------------------
+#
+# Here we will implement the full function reconstruction described in Sec. III A of [#GenPar]_.
+# We will show it both with equidistant and random shifts, corresponding to a uniform and a
+# non-uniform discrete Fourier transform (DFT), respectively.
```
Probably worth linking to the wikipedia articles here?

josh146 replied:
Also https://en.wikipedia.org/wiki/Trigonometric_interpolation? 😆 

In fact, I would probably expect to see a brief mention of this to introduce this section. That is, since `E(x)` is periodic, we can reconstruct it at a point _exactly_ by using the concept of trig interpolation

josh146 commented on demonstrations/tutorial_general_parshift.py:
```diff
@@ -0,0 +1,833 @@
+# Determining the full dependence on :math:`x`
+# --------------------------------------------
+#
+# Here we will implement the full function reconstruction described in Sec. III A of [#GenPar]_.
+# We will show it both with equidistant and random shifts, corresponding to a uniform and a
+# non-uniform discrete Fourier transform (DFT), respectively.
+# We start with the equidistant case, which is described in more detail in App. A2a and for which
+# we can directly implement Eq. (A15):
```
This is personal preference, but I tend to favour a more 'informal' style in demos, where explicit references (A2a, Sec III) are not provided in the main text but in a footnote, for readers interested in learning more. The main text is then more concerned with describing what is in the page, as opposed to relating it _directly_ to the structure of external content.

dwierichs replied:
That's a good call! :) I agree, in demos this is more distracting than useful.

josh146 commented on demonstrations/tutorial_general_parshift.py:
```diff
@@ -0,0 +1,833 @@
+# We start with the equidistant case, which is described in more detail in App. A2a and for which
+# we can directly implement Eq. (A15):
+#
+# .. math ::
+#
+#   x_\mu &= \frac{2\mu\pi}{2R+1}\\
+#   E(x) &=\sum_{\mu=-R}^R E\left(x_\mu\right) \frac{\sin\left(\frac{2R+1}{2}(x-x_\mu)\right)} {(2R+1)\sin \left(\frac{1}{2} (x-x_\mu)\right)}\\
+#   &=\sum_{\mu=-R}^R E\left(x_\mu\right) \frac{\operatorname{sinc}\left(\frac{2R+1}{2}(x-x_\mu)\right)} {\operatorname{sinc} \left(\frac{1}{2} (x-x_\mu)\right)},
```
I am tempted to say: rather than referring to an external equation A15, simply _replicate_ A15 directly here, briefly explain it, and _then_ apply it to the system at hand.

Otherwise, it might be conceptually difficult for an interested reader to follow the logical jumps in this section.

josh146 commented on demonstrations/tutorial_general_parshift.py:
```diff
@@ -0,0 +1,833 @@
+    ax.set_xlabel("$x$")
+    # Store the cost function for later
+    cost_functions.append(cost)
+_ = axs[0].set_ylabel("$E$")
```
These plots are so nice as well ⭐

mariaschuld replied:
I am wondering about the last line, why not `_ = axs[0].set_ylabel("$E$")` if you don't want to catch the return value?

josh146 replied:
```suggestion

_ = axs[0].set_ylabel("$E$")
```

dwierichs replied:
@mariaschuld you mean `axs[0].set_ylabel("$E$")`? I did this to avoid the return value being displayed in the demo.

josh146 replied:
@dwierichs  I realized I should have commented - I unresolved this comment thread since there was no response to my suggestion above, and I wasn't sure if it was missed.

I added the line break because the combination of the for loop indentation and `_` was messing with my eyes, and I wasn't clearly seeing the indentation level of this command 🙂

josh146 commented on demonstrations/tutorial_general_parshift.py:
```diff
@@ -0,0 +1,833 @@
+
+
+###############################################################################
+# *Works.*
```
🎉 💯 🦖

josh146 commented on demonstrations/tutorial_general_parshift.py:
```diff
@@ -0,0 +1,833 @@
+# further below.
+
+
+def compare_functions(originals, reconstructions, Ns, shifts, show_diff=True):
+    if show_diff:
+        fig, axs = plt.subplots(2, len(originals), figsize=(12, 4.5))
+    else:
+        fig, axs = plt.subplots(1, len(originals), figsize=(12, 2))
+    _axs = axs[0] if show_diff else axs
+    for i, (orig, recon, N, _shifts) in enumerate(zip(originals, reconstructions, Ns, shifts)):
+        E = np.array([orig(x) for x in X])
+        E_rec = np.array([recon(x) for x in X])
+        E_shifts = np.array([orig(shift) for shift in _shifts])
+
+        # Show E, the reconstruction, and the shifts (top)
+        _axs[i].plot(X, E, lw=2, color=orange)
+        _axs[i].plot(X, E_rec, linestyle=":", lw=3, color=green)
+        _axs[i].plot(_shifts, E_shifts, ls="", marker="x", c=red)
+        _axs[i].set_title(f"{N} qubits")
+        _axs[i].set_xticks([])
+        # Show the reconstruction error (bottom)
+        if show_diff:
+            E_shifts_rec = np.array([recon(shift) for shift in _shifts])
+            axs[1, i].plot(X, E - E_rec, color=blue)
+            axs[1, i].set_xlabel("$x$")
+    _ = _axs[0].set_ylabel("$E$")
+    if show_diff:
+        _ = axs[1, 0].set_ylabel("$E-E_{rec}$")
+    return axs
```
This codebloc could probably benefit from either:

- More extensive commenting, or
- Splitting it into separate reconstruction and plotting code blocks.

This is purely a pedagogical thought though! I realize it would probably lead to more and/or less efficient code 😆

dwierichs replied:
There is only plotting here :) The reconstructions already are an input as well. But yeah, I'll comment it a bit more and maybe add a comment that it's not important to understand this code in details, as it only plots..?

josh146 replied:
Ah, I think I was referring to 
```python
        E = np.array([orig(x) for x in X])
        E_rec = np.array([recon(x) for x in X])
        E_shifts = np.array([orig(shift) for shift in _shifts])
```

So maybe 'generation of the data' would have been a better description!

dwierichs replied:
I wrote a lot more comments now, I think in this case it would not only make the code less straight-forward but actually less understandable if we split up the evaluation and the plotting, because we would get a function 
```
cost_functions, reconstructions, **kwargs -> evaluation data
```
and a second function
```
evaluation data -> plots
```
leading to a lot of wrapping and unwrapping. I could imagine another approach where the plotting is done only for one function/reconstruction/subplot and then we use `map(compare_single_function)`. Do you think that would help?

josh146 replied:
I'll have a look at the current version at the next review and see how it flows 🙂

josh146 commented on demonstrations/tutorial_general_parshift.py:
```diff
@@ -0,0 +1,833 @@
+#
+# .. math ::
+#
+#   E(x_\mu) = a_0 + \sum_{\ell=1}^R a_{\ell}\cos(\ell x_\mu)+b_{\ell}\sin(\ell x_\mu)
```
I think this is the first explanation of `a_ell` and `b_ell`? I would definitely recommend having this appear earlier in the demo, since it is so core to the central tenet of _why_ this is possible.

josh146 commented on demonstrations/tutorial_general_parshift.py:
```diff
@@ -0,0 +1,833 @@
+
+def full_reconstruction_gen(fun, shifts):
+    """Reconstruct a univariate trigonometric function using arbitrary shifts."""
+    R = (len(shifts) - 1) // 2
+    frequencies = np.array(list(range(1, R + 1)))
+    # Construct the matrix C case by case
+    C1 = np.ones((2 * R + 1, 1))
+    C2 = np.cos(np.outer(shifts, frequencies))
+    C3 = np.sin(np.outer(shifts, frequencies))
+    C = np.hstack([C1, C2, C3])
+    # Evaluate the function to reconstruct at the shifted positions
+    evals = np.array([fun(shift) for shift in shifts])
+    # Solve the system of linear equations by inverting C
+    W = np.linalg.inv(C) @ evals
+    # Extract the Fourier coefficients
+    a0 = W[0]
+    a = W[1 : R + 1]
+    b = W[R + 1 :]
+    # Construct the Fourier series
+    def reconstruction(x):
+        return a0 + np.dot(a, np.cos(frequencies * x)) + np.dot(b, np.sin(frequencies * x))
```
recommend line breaks for readability 📚

josh146 commented on demonstrations/tutorial_general_parshift.py:
```diff
@@ -0,0 +1,833 @@
+# very close to each other, so that inverting the matrix :math:`C` becomes less stable numerically.
+# Conceptually, we see that the reconstruction does not rely on equidistant evaluations points.
+#
+# Too many sampling points and redundant reconstruction
+# -----------------------------------------------------
```
In terms of the flow of the demo, this section is probably not large enough to merit being a separate main section?

dwierichs replied:
Regarding the other parts, are there enough subsections/structure elements? I'm thinking whether removing this one or instead adding more elsewhere is better...

josh146 replied:
I'll need to take another look, but the ToC looked good to me on last review (except for this section!)

josh146 commented on demonstrations/tutorial_general_parshift.py:
```diff
@@ -0,0 +1,833 @@
+#
+# Now what can we do with these reconstruction methods? Before diving into the computation of
+# derivatives, a first idea is to obtain the global minimum of this univariate function.
+# This can be done via convex optimization (see Theorem 7 in [#CalcPQC]_), or with a global
+# optimization technique, which is feasible because we look at a function of a single parameter.
+# Two optimization techniques available via ``scipy.optimize`` are |brute|_ (a simple grid search
+# algorithm) and |shgo|_ (simplicial homology global optimization).
+# For this step we will use the equidistant shifts from the first reconstruction above.
+# Choose which global optimizer to use:
```
Seems like also a good place for wiki links :D

josh146 commented on demonstrations/tutorial_general_parshift.py:
```diff
@@ -0,0 +1,833 @@
+# paper, is *Rotosolve* and more generally, methods that optimize a subset of parameters at a
+# time are often referred to as *layerwise training*.
+# For completeness and as a minimal example, we here write out Rotosolve for quantum gates with
+# equidistant frequencies, but PennyLane also provides a full implementation via |Rotosolve_code|_.
```
Good to also link to the rotosolve demo!

this can be done using the following syntax:
```rest
Before going through this tutorial, we recommend that readers refer to the
:doc:`QNG tutorial </demos/tutorial_quantum_natural_gradient>` and
:doc:`VQE tutorial </demos/tutorial_vqe>` for overviews
of quantum natural gradient and the variational quantum eigensolver algorithm, respectively.
Let's get started!
```
(stole this from another demo)

josh146 commented on demonstrations/tutorial_general_parshift.py:
```diff
@@ -0,0 +1,833 @@
+# Generalized parameter-shift rules
+# ---------------------------------
+#
+# Next, we use a modified reconstruction strategy that only obtains the odd or even part of
+# :math:`E(x)` to compute odd- and even-order derivatives. While this method yields the entire
+# odd or even functional dependence in general (see below), we here directly make use of
+# Eqs. (26,27) in [#GenPar]_ to compute the first and second derivatives:
```
Same comment as above - I feel that all information required to read and conceptualize the demo should be directly in the body of the demo :)

josh146 commented on demonstrations/tutorial_general_parshift.py:
```diff
@@ -0,0 +1,833 @@
+# form as :math:`E(x)`.
+
+
+def parameter_shift(fun, R, order=1):
+    if order == 1:
```
I recommend splitting this up into two functions, (i.e., something like `param_shift` and `param_shift_second`), purely for readability

josh146 replied:
And restructing the text at the same time, e.g.:

- explicit param-shift intro/eqns
- param-shift code
- second order intro/eqns
- second-order code

dwierichs replied:
I left the comparison as a single block, though. Good?

josh146 commented on demonstrations/tutorial_general_parshift.py:
```diff
@@ -0,0 +1,833 @@
+    all_equal = (
+        "All entries match" if np.allclose(cost_grads, recon_grads) else "Some entries differ!"
+    )
+    print(f"{name} derivatives via jax: {all_equal}")
+    print("From the cost functions:       ", np.round(np.array(cost_grads), 5))
+    print(f"From the {recon_name} reconstructions: ", np.round(np.array(recon_grads), 5), "\n")
```
Amazing!!

josh146 replied:
Did you end up removing the jit from this notebook?

dwierichs replied:
Yes. This might have happened during an unrelated debugging process and might not have a reason at all :D Will look into it.

dwierichs replied:
Added `jax.jit`s where useful and possible

josh146 commented on demos_research.rst:
```diff
@@ -225,6 +225,12 @@ plateaus in quantum neural networks.
     :description: :doc:`demos/tutorial_quantum_analytic_descent`
     :tags: optimization model vqe
 
+.. customgalleryitem::
+    :tooltip: Reconstruct and differentiate univariate quantum functions.
+    :figure: demonstrations/general_parshift/thumbnail_genpar.png
+    :description: :doc:`demos/tutorial_general_parshift`
+    :tags: gradients reconstruction
```
Love the thumbnail image as well!

josh146 reviewed (changes requested):
Nice work @dwierichs! It's great to see all the results from the paper laid out like this, from start to finish alongside the numerics.

I've done a first-pass review, where I've mostly commented on the structure/flow, and haven't for the most part looked at grammar or code specifics.

Something that stands out to me after the first read: From the title and intro paragraph, the Rotosolve section definitely feels like a bit of a tangent. It adds to the length, and feels very much like a separate avenue of exploration separate to the main aim of the demo. In fact, 'Generalized Rotosolve' would probably work very well as a standalone demo.

In fact, it might also benefit from removing the 'too many sampling points' section, and potentially even rethinking the placement/length of the 'Automatically differentiated reconstructions' section.

licedric commented on demonstrations/tutorial_general_parshift.py:
```diff
@@ -0,0 +1,833 @@
+#
+# .. math ::
+#
+#   E'(0) &= \sum_{\substack{\mu=1\\\alpha\in\{+,-\}}}^R \alpha E\left(\alpha\frac{2\mu-1}{2R}\pi\right) \frac{(-1)^{\mu-1}}{4R\sin^2\left(\frac{2\mu-1}{4R}\pi\right)}, \\
```
We didn't do this in our paper, but the formula can be simplified by removing the sum over alpha (the sign) and let mu run from 1 to 2R.

licedric replied:
And of course if you make this change this simplifies the code below as well.

dwierichs replied:
I like this simplification a lot. For the first derivative $E'(0)$, I am not sure how to include the sign $\alpha$ in front of the coefficients best.

dwierichs replied:
Nevermind, found that it is absorbed in the $\pi$ shift anyways :)

licedric commented on demonstrations/tutorial_general_parshift.py:
```diff
@@ -0,0 +1,833 @@
+# .. math ::
+#
+#   E'(0) &= \sum_{\substack{\mu=1\\\alpha\in\{+,-\}}}^R \alpha E\left(\alpha\frac{2\mu-1}{2R}\pi\right) \frac{(-1)^{\mu-1}}{4R\sin^2\left(\frac{2\mu-1}{4R}\pi\right)}, \\
+#   E''(0) &= -E(0)\frac{2R^2+1}{6} + E(\pi)\frac{(-1)^{R-1}}{2}+ \sum_{\substack{\mu=1\\\alpha\in\{+,-\}}}^{R-1} E\left(\alpha\frac{\mu\pi}{R}\right)\frac{(-1)^{\mu-1}}{2\sin^2 \left(\frac{\mu\pi}{2R}\right)}.
```
Ditto here: we can simplify the formula by removing the E(pi) special case term, removing the sum over alpha, and letting mu run from 1 to 2R-1. We still need the E(0) term though.

dwierichs replied:
This is nice!

mariaschuld commented:
> @mariaschuld Would you be interested in reviewing this next week?

Sure, preferably after the first iterations are done?

josh146 commented on demonstrations/tutorial_general_parshift.py:
```diff
@@ -17,22 +17,34 @@
-Fourier transforms. In the case of general shifts for the shift rules, linear algebra actually
-is all we use for the reconstruction.
-We briefly recap the derivation of the functional form of :math:`E(x)` further below, the full
+Before going through this tutorial, we recommend that readers refer to the
+:doc:`Rotosolve & Rotoselect tutorial </demos/tutorial_rotoselect>`, and having a basic
+understanding of the :doc:`parameter-shift rule </glossary/parameter_shift>` might make
+this tutorial easier to dive into.
```
Perhaps either?

- Turn this into a `.. note::`, so that it doesn't look like the 'opening' to the demo
- Or turn into a note and move to the end of the introduction

josh146 commented on demonstrations/tutorial_general_parshift.py:
```diff
@@ -17,22 +17,34 @@
+understanding of the :doc:`parameter-shift rule </glossary/parameter_shift>` might make
+this tutorial easier to dive into.
+
+In this demo we will learn how to reconstruct univariate quantum functions, i.e. those that
+depend on a single parameter. This reconstruction and variants thereof allow us to derive more
```
```suggestion
In this demo we will learn how to reconstruct univariate quantum functions, i.e., those that
depend on a single parameter. This reconstruction and variants thereof allow us to derive more
```

josh146 replied:
I'll wait for @mariaschuld's thoughts because I'm too deep into this 😆 

But my gut feeling is that this might be too technical for an opening.

dwierichs replied:
I think I agree. I'm happy for suggestions here but will also try to simplify it further :)

josh146 commented on demonstrations/tutorial_general_parshift.py:
```diff
@@ -17,22 +17,34 @@
+`Analytic gradients in variational quantum algorithms: Algebraic extensions of the parameter-shift rule to general unitary transformations <https://arxiv.org/abs/2107.08131>`_ [#AlgeShift]_ and
+one focusing on special gates and spectral decompositions, namely
+`Generalized quantum circuit differentiation rules <https://arxiv.org/abs/2108.01218>`_
+[#GenDiffRules]_.
```
Since this is a demo presenting _your_ work, I am tempted to say that these other papers might not need to be so explicitly called out in the intro 🤔 But curious what others think. Perhaps they are better suited for the conclusion or a footnote.

dwierichs replied:
Happy to hear Maria's thought on this, I'm kinda torn :D

josh146 commented on demonstrations/tutorial_general_parshift.py:
```diff
@@ -112,56 +127,85 @@ def random_observable(N, seed):
 
 ###############################################################################
-# Section II A of [#GenPar]_ shows that the output of this circuit, namely the function
+# Let's also prepare some plotting functionalities and colors:
```
💯 this makes it much easier to read!

josh146 commented on demonstrations/tutorial_general_parshift.py:
```diff
@@ -547,74 +639,76 @@ def cost(param):
+#
+# .. math ::
+#
+#   E''(0) &= -E(0)\frac{2R^2+1}{6} - \sum_{\mu=1}^{2R-1} E\left(\frac{\mu\pi}{R}\right)\frac{(-1)^\mu}{2\sin^2 \left(\frac{\mu\pi}{2R}\right)}.
```
This expression doesn't seem to render 🤔

josh146 commented on demonstrations/tutorial_general_parshift.py:
```diff
@@ -547,74 +639,76 @@ def cost(param):
 ###############################################################################
 # Let's compare these two shift rules to the finite-difference derivative commonly used for
-# numerical differentiation. We choose a shift value of :math:`d_x=5\cdot10^{-5}`.
+# numerical differentiation. We choose a finite difference of :math:`d_x=5\cdot10^{-5}`.
```
```suggestion
# numerical differentiation. We choose a finite difference of :math:`d_x=5\times 10^{-5}`.
```

dwierichs commented:
@josh146 I think there are only a few quite minor things open from the first round of reviews, on which I'm looking forward to your feedback on the changes I made there.
Regarding the major structure change of skipping (some of) the sections "Rotosolve", "too many sampling points" and "Automatically differentiated reconstructions", I did not take action yet and would combine this with the feedback by Maria.
@mariaschuld , if you have time I would love to hear your thought on this current version :) There are two particular aspects open in the review by Josh above.

mariaschuld commented on demonstrations/tutorial_general_parshift.py:
```diff
@@ -0,0 +1,985 @@
+All we will need for the demo is the insight that these functions are Fourier series in their
+variable, and a bit of knowledge about Fourier transforms.
+In the most general case, the shift rules become even simpler conceptually (but are a bit more
+computation-intensive).
```
After reading the first paragraph I don't know what a reconstruction is, and so I don't really have the tools to access if the demo is important for me. Maybe add a little equation? And make the "what it is important for" more concrete with links to papers, demos or more simple equations?

mariaschuld commented on demonstrations/tutorial_general_parshift.py:
```diff
@@ -0,0 +1,985 @@
+In the most general case, the shift rules become even simpler conceptually (but are a bit more
+computation-intensive).
+
+We will briefly recap the derivation of the functional form below, the full
```
What is a functional form here?

mariaschuld commented on demonstrations/tutorial_general_parshift.py:
```diff
@@ -0,0 +1,985 @@
+`Analytic gradients in variational quantum algorithms: Algebraic extensions of the parameter-shift rule to general unitary transformations <https://arxiv.org/abs/2107.08131>`_ [#AlgeShift]_ and
+one focusing on special gates and spectral decompositions, namely
+`Generalized quantum circuit differentiation rules <https://arxiv.org/abs/2108.01218>`_
+[#GenDiffRules]_.
```
Can this be simplified? The first time the observation was made was in X, we follow Y, but there are also other relevant refs Z?

mariaschuld commented on demonstrations/tutorial_general_parshift.py:
```diff
@@ -0,0 +1,985 @@
+    Before going through this tutorial, we recommend that readers refer to the
+    :doc:`Rotosolve & Rotoselect tutorial </demos/tutorial_rotoselect>`, and having a basic
+    understanding of the :doc:`parameter-shift rule </glossary/parameter_shift>` might make
+    this tutorial easier to dive into.
```
It's quite hard to see for a non-expert how the Rotosolve tutorial links to the insights here. Maybe make this a "further reading" suggestion and mention _what_ exactly from rotosolve applies here?

mariaschuld commented on demonstrations/tutorial_general_parshift.py:
```diff
@@ -0,0 +1,985 @@
+parameter.
+For this we will use a certain gate structure that allows us to tune the number of eigenvalues
+of the parametrized unitary, and thus the number of frequencies in the cost function with respect
+to the variational parameter.
```
At this stage the entire links between eigenvalues and frequencies are not known to the reader and this sentence is more scary than enlightning? 

Maybe just explain the circuit structure and refer to its use a bit further below? Or explain the equations first and then define the circuit?

mariaschuld commented on demonstrations/tutorial_general_parshift.py:
```diff
@@ -0,0 +1,985 @@
+#
+# .. math ::
+#
+#   E(x) = a_0 + \sum_{\ell=1}^R a_{\ell}\cos(\ell x)+b_{\ell}\sin(\ell x).
```
This is the first time the reader really gets to the core of this whole Fourier series business. I would have this equation in the intro, or at the top of this section?

mariaschuld commented on demonstrations/tutorial_general_parshift.py:
```diff
@@ -0,0 +1,985 @@
+# real-valued. This is because :math:`B` is Hermitian, which implies (anti-)symmetry for the
+# real (imaginary) Fourier coefficients.
+#
+# This is true for any number of qubits (and therefore ``RZ`` gates) we use.
```
This sentence belongs to the paragraph before, right?

mariaschuld commented on demonstrations/tutorial_general_parshift.py:
```diff
@@ -0,0 +1,985 @@
+
+###############################################################################
+# Indeed we see that :math:`E(x)` is a periodic function whose complexity grows when increasing
+# the number of gates parametrized by :math:`x`.
```
That's a nice description! Isn't that what you want to mention above when you motivate this circuit?

mariaschuld commented on demonstrations/tutorial_general_parshift.py:
```diff
@@ -0,0 +1,985 @@
+generator = 0.5 * np.sum(np.array(list(map(PauliZ, range(N)))), axis=0)
+# Show the unique eigenvalues
+omegas = sorted(list(set(generator.astype(int))))
+print(f"The eigenvalues are: {omegas}")
```
Just FYI, maybe one could also use the `fourier.spectrum` function here

mariaschuld commented on demonstrations/tutorial_general_parshift.py:
```diff
@@ -0,0 +1,985 @@
+        )
+    )
+)
+print(f"The frequencies are: {Omegas}")
```
I think almost all of the above is explained in the Fourier series demo. Wouldn't this be a much easier pre-read recommendation?

mariaschuld commented on demonstrations/tutorial_general_parshift.py:
```diff
@@ -0,0 +1,985 @@
+            )
+        )
+    )
+)
```
Uh, I think for the sake of readability one needs to do this in a few more lines here...

mariaschuld commented on demonstrations/tutorial_general_parshift.py:
```diff
@@ -0,0 +1,985 @@
+# Determining the full dependence on :math:`x`
+# --------------------------------------------
+#
+# Here we will implement the full function reconstruction of :math:`E(x)`.
```
Still don't know what this is :) Does this mean you want to find a classical description that approximates `E(x)` to a certain value?

mariaschuld commented on demonstrations/tutorial_general_parshift.py:
```diff
@@ -0,0 +1,985 @@
+
+reconstructions_equ = list(map(full_reconstruction_equ, cost_functions, Ns))
+equ_shifts = [[2 * mu * np.pi / (2 * N + 1) for mu in range(-N, N + 1)] for N in Ns]
+axs = compare_functions(cost_functions, reconstructions_equ, Ns, equ_shifts)
```
Very nice plots!

mariaschuld commented on demonstrations/tutorial_general_parshift.py:
```diff
@@ -0,0 +1,985 @@
+
+
+###############################################################################
+# *Works.*
```
I wasn't sure if this is a rendering issue :) would it better be "It works!"?

mariaschuld commented on demonstrations/tutorial_general_parshift.py:
```diff
@@ -0,0 +1,985 @@
+# highly relevant in applications.
+# In addition, not only the number of distinct circuits to evaluate grows with :math:`R`, which is
+# important for simulators and cloud queueing, but also the number of required measurements, which
+# is meaningful for the (time) complexity of quantum operations, does so as well!
```
If you want to shorten, I feel this last experiment is very specific and could just be a comment!

dwierichs replied:
Good idea! :)

mariaschuld commented on demonstrations/tutorial_general_parshift.py:
```diff
@@ -0,0 +1,985 @@
+#
+# Rotosolve
+# ---------
+#
```
I would link the demo here, not above!

mariaschuld commented on demonstrations/tutorial_general_parshift.py:
```diff
@@ -0,0 +1,985 @@
+
+from scipy.optimize import brute, shgo
+
+optimizer = "brute"  # "shgo"
```
Maybe say `# can be changed to "shgo"`

mariaschuld commented on demonstrations/tutorial_general_parshift.py:
```diff
@@ -0,0 +1,985 @@
+# The idea of using this one-dimensional minimization for functions of multiple variables, by
+# applying it coordinate-wise, was proposed in general in [#CalcPQC]_ and for Pauli rotation
+# gates in particular in [#Rotosolve]_. The name for this algorithm, put forward by the latter
+# paper, is *Rotosolve* and more generally, methods that optimize a subset of parameters at a
```
```suggestion
# paper, is *Rotosolve* and methods that optimize a subset of parameters at a
```
repetition of "general"

mariaschuld commented on demonstrations/tutorial_general_parshift.py:
```diff
@@ -0,0 +1,985 @@
+# gates in particular in [#Rotosolve]_. The name for this algorithm, put forward by the latter
+# paper, is *Rotosolve* and more generally, methods that optimize a subset of parameters at a
+# time are often referred to as *layerwise training*.
+# For completeness and as a minimal example, we here write out Rotosolve for quantum gates with
```
Mention that the reason to do this is that the example shows how to integrate the reconstruction explained above?

mariaschuld reviewed (changes requested):
Great and very detailed demo @dwierichs!

I still think there needs to be a bit of polishing to make it more accessible. At the moment, I don't know until very very late what is actually expecting me - the content is rather linear and a lot of details happen before the core idea is there. 

Is there a way to make it more easy-to-specific? Like, in the introduction to summarise what a reconstruction is, and list the use cases that you explore, explaining where the reconstruction helps in each one. Readers that only want to understand the gist (which is probably 95% of them) should have a chance to spend 10min and get it!

I think these will be minor changes with a huge impact. The content is great, especially theplots!

dwierichs commented:
After having presented a similar subset of the results in a discussion/talk, I think restructured way that is more accessible, following both, @mariaschuld 's idea of simplifying things and going from easy to specific and @josh146 's suggestion to skip Rotosolve for this demo.
1. The demo now introduces the functional form in a more straight-forwards approach and leaves the technicalities mostly to the Fourier demo, as suggested by Maria, this now also is the main reference for read-up before looking at this demo.
2. Rotosolve and the "overparametrized" reconstruction are gone (the latter being a note-box now).
3. `fourier.spectrum` is used instead of manually looking at the generator and the frequencies from its eigenvalues - in line with skipping the technical derivation.
4. The general parameter shift rule now comes _after_ the partial reconstruction of odd/even parts only. This makes much more sense as they are a manual replacement for the autodiff approach which can be understood better after having seen the partial reconstruction idea, imho. This puts the namesake of the demo at the very end, I guess that's fine, though (?)
5. All of Maria's comments should be addressed in the changes I made.

I'm sorry for this major overthrowing of some parts and the additional effort the review requires because of that, but I really hope this restructuring helped the quality a good bit :)

mariaschuld commented on demonstrations/tutorial_general_parshift.py:
```diff
@@ -0,0 +1,837 @@
+    :doc:`Fourier series expressiveness tutorial </demos/tutorial_expressivity_fourier_series>`.
+    Additionally, having a basic understanding of the
+    :doc:`parameter-shift rule </glossary/parameter_shift>` might make this tutorial easier 
+    to dive into.
```
The beginning is MUCH better! :100:

josh146 replied:
Agree! The intro is 💯 It gives a great overview of what to expect

mariaschuld commented on demonstrations/tutorial_general_parshift.py:
```diff
@@ -0,0 +1,837 @@
+
+Cost functions arising from quantum gates
+-----------------------------------------
+Let's start our investigations with a cost function that arises from a single quantum operation
+which is parametrized by a single variational parameter :math:`x`.
```
I don't know how to best do it, but would it be clearer to say that you are interpreting expectations as the cost function here? I would find that an extremely important limitation and clarification of what we are trying to do!

dwierichs replied:
Gave it a shot.

mariaschuld commented on demonstrations/tutorial_general_parshift.py:
```diff
@@ -0,0 +1,837 @@
+
+Let's implement such a cost function using PennyLane.
+We begin with functions that generate the random initial state :math:`|\psi\rangle`
+and the random observable :math:`B` for a given number of qubits :math:`N` and a fixed seed:
```
Why do you need randomness here? Or is it just to make no assumptions on those two objects?

mariaschuld replied:
I am asking because you spend the second most important bit of the demo, the first lines of code a user sees, just on "how to create and validate random observables/states". That's quite a price to pay if it is just there to show people how it's done (although, I find it quite useful to see ;) - just not when I want to learn about other things?)

dwierichs replied:
Yes, exactly. This is actually was easier for me than coming up with a short circuit for arbitrary qubit numbers that actually creates a cost function that is as complex as it could be for that operation. Is this confusing? Then I could work out explicit circuits. 
I thought this might also follow the mathy part nicely because we don't talk about $\psi$ or $B$ - and we don't need to :)

dwierichs replied:
Ah sorry, your second comment hadn't arrived when I loaded the page - I'll know better next time :D
You're right, this is quite highlighted the way it is and it's nothing essential to this demo.
Two ideas:
1. Skip the validation of that these utilities do what we want and thus shorten the distraction.
2. Move the definition _after_ `make_cost` a la "Oh, but wait, we have not defined what `random_state` is or does. Here it is: ... (This would be possible because the two utilities are not called until the 5th code block).

mariaschuld commented on demonstrations/tutorial_general_parshift.py:
```diff
@@ -0,0 +1,837 @@
+
+.. math ::
+
+  E(x)=\langle\psi | U^\dagger(x) B U(x)|\psi\rangle
```
Just out of curiosity, would it not be more intuitive to name this `C(x)` or even `Cost(x)`? Ignore if this is just following the paper!

dwierichs replied:
Yeah, it's following the paper :sweat_smile:

josh146 replied:
```suggestion
  E(x)=\langle\psi | U^\dagger(x) B U(x)|\psi\rangle.
```

mariaschuld commented on demonstrations/tutorial_general_parshift.py:
```diff
@@ -0,0 +1,837 @@
+
+  U(x)=\prod_{a=1}^N R_Z^{(a)}(x) = \prod_{a=1}^N \exp\left(-i\frac{x}{2} Z_a\right).
+
+Let's implement such a cost function using PennyLane.
```
Nitpicking: maybe remove half of the "Let's" and use some other phrase? :)

dwierichs replied:
Thanks, very good!

mariaschuld commented on demonstrations/tutorial_general_parshift.py:
```diff
@@ -0,0 +1,837 @@
+
+###############################################################################
+# Now that we took care of these preparations, let's dive right into it:
+# It can rather easily be shown [#GenPar]_ that :math:`E(x)` takes the form of a
```
I would find it useful to see the Fourier module link repeated here, to say "if that is not clear to you, this is why you should read there"

dwierichs replied:
You mean the demo link or the module link? :) I agree though :)

dwierichs replied:
Just added both :)

mariaschuld commented on demonstrations/tutorial_general_parshift.py:
```diff
@@ -0,0 +1,837 @@
+# i.e., we will *reconstruct* :math:`E(x)`.
+# The key idea is not new: Since :math:`E(x)` is periodic with known, integer frequencies, we can
+# reconstruct it *exactly* by using
+# `trigonometric interpolation <https://en.wikipedia.org/wiki/Trigonometric_interpolation>`_.
```
I think I would have liked to see this keyword in the introduction, because this is the core of the method you are using!

dwierichs replied:
This is actually a better phrasing of "a bit of knowledge about Fourier transforms" in the intro, because that's essentially the "bit". I changed this in the intro, moving the wiki link to the intro as well.

mariaschuld commented on demonstrations/tutorial_general_parshift.py:
```diff
@@ -0,0 +1,837 @@
+#   E(x) &=\sum_{\mu=-R}^R E\left(x_\mu\right) \frac{\sin\left(\frac{2R+1}{2}(x-x_\mu)\right)} {(2R+1)\sin \left(\frac{1}{2} (x-x_\mu)\right)}\\
+#
+# where we reformulated :math:`E` in the second expression using the sinc function
+# () to enhance the numerical stability.
```
Are the `()` a typo?

dwierichs replied:
Oh, yes sorry.

mariaschuld commented on demonstrations/tutorial_general_parshift.py:
```diff
@@ -0,0 +1,837 @@
+#     :math:`\operatorname{sinc}(0)=1` so that the denominator does no longer vanish at
+#     the shifted points. Note that we have to take care of a rescaling factor of :math:`\pi`
+#     between this definition of :math:`\operatorname{sinc}` and the NumPy implementation
+#     ``np.sinc``.
```
That's very useful, and well-posed as a note!

josh146 replied:
How does the NumPy implementation differ?

mariaschuld commented on demonstrations/tutorial_general_parshift.py:
```diff
@@ -0,0 +1,837 @@
+# We may either perform the partial reconstruction and defer the differentiation to an
+# autodifferentiation framework like JAX, or compute the derivative of the (partial) 
+# reconstructions manually and directly implement the resulting shift rule. 
+# We start with the former:
```
This is tough to understand for anyone but experts, but also hard to make easier to digest...

dwierichs replied:
I'll give it another shot! :)

josh146 replied:
```suggestion
# We start with the former.
```

mariaschuld commented on demonstrations/tutorial_general_parshift.py:
```diff
@@ -0,0 +1,837 @@
+# of interest. For odd-order derivatives, we use the reconstruction of the odd part, for the
+# even-order derivatives that of the even part.
+#
+# As mentioned above, we need to modify the Dirichlet kernels for this, and we will use equidistant 
```
This may be a leftover from before, but I don't think you introduce Dirichlet kernels!

josh146 replied:
Was just about to say the same :)

josh146 replied:
It was introduced in a code comment above, but in an off-hand manner, and not in the text.

mariaschuld commented on demonstrations/tutorial_general_parshift.py:
```diff
@@ -0,0 +1,837 @@
+
+
+###############################################################################
+# Great! The even and odd part indeed combine into the correct function again. But what did we
```
I have to admit, I was struggling a bit to see this. Do you mean their sum or absolute product? Can you give the reader a hint how to "see" this?

dwierichs replied:
Mhm.. I kinda agree it's not so easy to see. It's actually plotted! The dashed line is dashed, because it is the original function (green) and the sum of the red and blue parts on top of it (orange). Maybe a legend and writing this in addition to 

> Let’s now look at these even (blue) and odd (red) reconstructions and how they indeed combine to the full function (we will use the compare_functions utility from above for the latter).

would help? :)

josh146 replied:
I definitely second a legend!

josh146 replied:
I think a visual cue is more helpful than additional textual cues

mariaschuld commented on demonstrations/tutorial_general_parshift.py:
```diff
@@ -0,0 +1,837 @@
+print(f"Second-order finite difference:    {np.round(np.array(fd_der2), 6)}")
+
+###############################################################################
+# Great, the parameter-shift rules work as expected! And we were able to save
```
Maybe remove one or two "great" :)

mariaschuld commented on demonstrations/tutorial_general_parshift.py:
```diff
@@ -0,0 +1,837 @@
+Once we have this reconstruction, we use it to compute analytically exact derivatives
+of the quantum function. We implement this once by using autodifferentiation on the classical
+function and once by computing the derivative manually, resulting in generalized parameter-shift
+rules for quantum functions.
```
These two methods require expert knowledge to really understand. I wonder if you can tell the reader more about their role, for example which one should be used in practice, and why you show both!

josh146 replied:
Perhaps one paragraph per method?

dwierichs replied:
I expanded this slightly, paraphrasing the new, bigger comparison at the end in shortened form.

mariaschuld reviewed (changes requested):
@dwierichs well done on condensing this, it is really a lot more accessible! 
I don't have major blockers, only a few suggestions. The main one is to tell the reader a bit more about the two methods if you can - I think 95% of your readers will not have the background knowledge to really understand what they mean. But that is not a problem, as long as you explain why they are important, what role they play, which one is better than the others, or why you introduce both - this gives the reader the means to judge whether she should invest into understanding the details.

I was wondering about the Binary file deletions? They seem unrelated with your tutorial...

josh146 commented on demonstrations/tutorial_general_parshift.py:
```diff
@@ -0,0 +1,837 @@
+
+
+# Figure with multiple axes
+fig, axs = plt.subplots(1, len(Ns), figsize=(12, 2))
+for ax, N, E in zip(axs, Ns, evaluated_cost):
```
```suggestion
fig, axs = plt.subplots(1, len(Ns), figsize=(12, 2))

for ax, N, E in zip(axs, Ns, evaluated_cost):
```

josh146 commented on demonstrations/tutorial_general_parshift.py:
```diff
@@ -0,0 +1,837 @@
+
+
+###############################################################################
+# Indeed we see that :math:`E(x)` is a periodic function whose complexity grows when increasing
+# :math:`N` and therefore the number of gates parametrized by :math:`x`.
```
Something in this sentence doesn't sound right to me 🤔 Particularly the "and therefore the number of gates parametrized by :math:`x`." part (the implication is not clear)

josh146 commented on demonstrations/tutorial_general_parshift.py:
```diff
@@ -0,0 +1,837 @@
+
+
+###############################################################################
+# Indeed we see that :math:`E(x)` is a periodic function whose complexity grows when increasing
```
```suggestion
#
# |
#
# Indeed we see that :math:`E(x)` is a periodic function whose complexity grows when increasing
```

josh146 replied:
This inserts a line break, since the text is a bit close to the figure!

dwierichs replied:
Ah, that's great!

josh146 commented on demonstrations/tutorial_general_parshift.py:
```diff
@@ -0,0 +1,837 @@
+# 
+# .. note ::
+# 
+#     The analysis tool :func:`~.pennylane.fourier.spectrum` makes use of the internal
+#     structure of the :class:`~.pennylane.QNode` that encodes the cost function.
+#     As we used the ``jax.jit`` decorator when defining the cost function above, we
+#     here need to pass the wrapped function to ``spectrum``, which is stored in
+#     ``cost_function.__wrapped__``.
```
nice catch/solution

josh146 commented on demonstrations/tutorial_general_parshift.py:
```diff
@@ -0,0 +1,837 @@
+#   E(x) = a_0 + \sum_{\ell=1}^R a_{\ell}\cos(\ell x)+b_{\ell}\sin(\ell x).
+#
+# Here, :math:`a_{\ell}` and :math:`b_{\ell}` are the *Fourier coefficients*.
+# Due to :math:`B` being Hermitian, :math:`E(x)` is a real-valued function so that
+# only positive frequencies and real coefficients appear in the Fourier series for :math:`E(x)`.
```
```suggestion
# Due to :math:`B` being Hermitian, :math:`E(x)` is a real-valued function, so
# only positive frequencies and real coefficients appear in the Fourier series for :math:`E(x)`.
```

josh146 commented on demonstrations/tutorial_general_parshift.py:
```diff
@@ -0,0 +1,837 @@
+# the zero-frequency contribution in the coefficient :math:`a_0`.
+# If you are interested why the number of gates coincides with the number of frequencies,
+# check out the :doc:`Fourier module tutorial </demos/tutorial_expressivity_fourier_series>`.
+# Before moving on, let's also have a look at the Fourier coefficients in the functions
+# we created:
```
```suggestion
#
# Before moving on, let's also have a look at the Fourier coefficients in the functions
# we created:
```

josh146 commented on demonstrations/tutorial_general_parshift.py:
```diff
@@ -0,0 +1,837 @@
+
+###############################################################################
+# We find the real (imaginary) Fourier coefficients to be (anti-)symmetric.
+# This is expected because :math:`E(x)` is real-valued and we again see why it is enough
+# to consider positive frequencies: The coefficients of the negative frequencies follow
+# from those of the positive frequencies.
```
```suggestion
# This is expected because :math:`E(x)` is real-valued, and we again see why it is enough
# to consider positive frequencies: the coefficients of the negative frequencies follow
# from those of the positive frequencies.
```

josh146 commented on demonstrations/tutorial_general_parshift.py:
```diff
@@ -0,0 +1,837 @@
+# `uniform <https://en.wikipedia.org/wiki/Discrete_Fourier_transform>`_ and a
+# `non-uniform <https://en.wikipedia.org/wiki/Non-uniform_discrete_Fourier_transform>`_
+# discrete Fourier transform (DFT), respectively.
+# We start with the equidistant case for which we can directly implement the trigonometric
+# interpolation:
```
```suggestion
#
# We start with the equidistant case for which we can directly implement the trigonometric
# interpolation:
```

josh146 commented on demonstrations/tutorial_general_parshift.py:
```diff
@@ -0,0 +1,837 @@
+# .. math ::
+#
+#   x_\mu &= \frac{2\mu\pi}{2R+1}\\
+#   E(x) &=\sum_{\mu=-R}^R E\left(x_\mu\right) \frac{\sin\left(\frac{2R+1}{2}(x-x_\mu)\right)} {(2R+1)\sin \left(\frac{1}{2} (x-x_\mu)\right)}\\
```
```suggestion
#   E(x) &=\sum_{\mu=-R}^R E\left(x_\mu\right) \frac{\sin\left(\frac{2R+1}{2}(x-x_\mu)\right)} {(2R+1)\sin \left(\frac{1}{2} (x-x_\mu)\right)},\\
```

josh146 commented on demonstrations/tutorial_general_parshift.py:
```diff
@@ -0,0 +1,837 @@
+#   x_\mu &= \frac{2\mu\pi}{2R+1}\\
+#   E(x) &=\sum_{\mu=-R}^R E\left(x_\mu\right) \frac{\sin\left(\frac{2R+1}{2}(x-x_\mu)\right)} {(2R+1)\sin \left(\frac{1}{2} (x-x_\mu)\right)}\\
+#
+# where we reformulated :math:`E` in the second expression using the sinc function
+# () to enhance the numerical stability.
```
```suggestion
# where we reformulated :math:`E` in the second expression using the
`sinc function <https://en.wikipedia.org/wiki/Sinc_function>`__ to enhance the numerical stability.
```

josh146 commented on demonstrations/tutorial_general_parshift.py:
```diff
@@ -0,0 +1,837 @@
+#
+# .. note ::
+#
+#     When implementing :math:`E`, we will replace 
+#     :math:`\frac{\sin\left(\frac{2R+1}{2}(x-x_\mu)\right)} {(2R+1)\sin \left(\frac{1}{2} (x-x_\mu)\right)}`
+#     by :math:`\frac{\operatorname{sinc}\left(\frac{2R+1}{2}(x-x_\mu)\right)} {\operatorname{sinc} \left(\frac{1}{2} (x-x_\mu)\right)}` where the sinc function is defined as 
+#     :math:`\operatorname{sinc}(x)=\sin(x)/x`. This enhances the numerical stability because
```
Since these equations are quite large, suggest moving to display mode :)

josh146 replied:
```suggestion
#     When implementing :math:`E`, we will replace 
#     :math:`\frac{\sin\left(\frac{2R+1}{2}(x-x_\mu)\right)} {(2R+1)\sin \left(\frac{1}{2} (x-x_\mu)\right)}`
#     by :math:`\frac{\operatorname{sinc}\left(\frac{2R+1}{2}(x-x_\mu)\right)} {\operatorname{sinc} \left(\frac{1}{2} (x-x_\mu)\right)}` where the sinc function is defined as 
#     :math:`\operatorname{sinc}(x)=\sin(x)/x`. This enhances the numerical stability since
```

josh146 commented on demonstrations/tutorial_general_parshift.py:
```diff
@@ -0,0 +1,837 @@
+#     :math:`\frac{\sin\left(\frac{2R+1}{2}(x-x_\mu)\right)} {(2R+1)\sin \left(\frac{1}{2} (x-x_\mu)\right)}`
+#     by :math:`\frac{\operatorname{sinc}\left(\frac{2R+1}{2}(x-x_\mu)\right)} {\operatorname{sinc} \left(\frac{1}{2} (x-x_\mu)\right)}` where the sinc function is defined as 
+#     :math:`\operatorname{sinc}(x)=\sin(x)/x`. This enhances the numerical stability because
+#     :math:`\operatorname{sinc}(0)=1` so that the denominator does no longer vanish at
```
Should this technically be a limit?
```suggestion
#     :math:`\operatorname{sinc}(0)=1`, so that the denominator does no longer vanish at
```

josh146 commented on demonstrations/tutorial_general_parshift.py:
```diff
@@ -0,0 +1,837 @@
+
+
+###############################################################################
+# *It Works!*
```
```suggestion
# *It works!*
```

josh146 commented on demonstrations/tutorial_general_parshift.py:
```diff
@@ -0,0 +1,837 @@
+# *It Works!*
+#
+# Now let's test the reconstruction with less regular sampling points on which to evaluate
+# :math:`E`. This means we can no longer use the closed-form expression from above but switch
+# to solving the set of equations
```
```suggestion
# :math:`E`. This means we can no longer use the closed-form expression from above, but switch
# to solving the set of equations
```

josh146 commented on demonstrations/tutorial_general_parshift.py:
```diff
@@ -0,0 +1,837 @@
+#
+#   E(x_\mu) = a_0 + \sum_{\ell=1}^R a_{\ell}\cos(\ell x_\mu)+b_{\ell}\sin(\ell x_\mu)
+#
+# with the --- now irregular --- sampling points :math:`x_\mu`.
```
```suggestion
# with the---now irregular---sampling points :math:`x_\mu`.
```

josh146 commented on demonstrations/tutorial_general_parshift.py:
```diff
@@ -0,0 +1,837 @@
+###############################################################################
+# *It Works!*
+#
+# Now let's test the reconstruction with less regular sampling points on which to evaluate
```
I wonder if it is worth having a subsection heading here? "Non-equidistant shifts"?

Mainly to emphasize that this is an another method of getting the same result

dwierichs replied:
Yes, I like that! I also added one for the equidistant shifts, for consistency

josh146 commented on demonstrations/tutorial_general_parshift.py:
```diff
@@ -0,0 +1,837 @@
+#   E_\text{even}(x) &= \sum_{\mu=0}^R E_\text{even}(x_\mu) \hat{D}_\mu(x)\\
+#   \hat{D}_\mu(x) &=
+#   \begin{cases}
+#      \frac{\sin(Rx)}{2R \tan(x/2)} &\text{if } \mu = 0 \\
+#      \frac{\sin(R (x-x_\mu))}{2R \tan\left(\frac{1}{2} (x-x_\mu)\right)} + \frac{\sin(R (x+x_\mu))}{2R \tan\left(\frac{1}{2} (x+x_\mu)\right)} & \text{if } \mu \in [R-1] \\
+#      \frac{\sin(R (x-\pi))}{2R \tan\left(\frac{1}{2} (x-\pi)\right)} & \text{if } \mu = R.
```
```suggestion
#      \frac{\sin(Rx)}{2R \tan(x/2)} &\text{if } \mu = 0 \\[12pt]
#      \frac{\sin(R (x-x_\mu))}{2R \tan\left(\frac{1}{2} (x-x_\mu)\right)} + \frac{\sin(R (x+x_\mu))}{2R \tan\left(\frac{1}{2} (x+x_\mu)\right)} & \text{if } \mu \in [R-1] \\[12pt]
#      \frac{\sin(R (x-\pi))}{2R \tan\left(\frac{1}{2} (x-\pi)\right)} & \text{if } \mu = R.
```
larger line spacing might help here!

dwierichs replied:
Such nice tricks! :)

josh146 commented on demonstrations/tutorial_general_parshift.py:
```diff
@@ -0,0 +1,837 @@
+
+###############################################################################
+# Great! The even and odd part indeed combine into the correct function again. But what did we
+# gain? Nothing, actually, for the full reconstruction! Quite the opposite, we spent :math:`2R`
+# evaluations of :math:`E` on each part, that is :math:`4R` evaluations overall to obtain a
+# description of the full function :math:`E`! This is way more than the :math:`2R+1` 
+# evaluations needed for the full reconstructions from the beginning.
+# However, remember that we set out to compute derivatives of :math:`E` at :math:`0`, so that
+# for derivatives of odd/even order only the odd/even reconstruction is required.
+# Using an autodifferentiation framework, e.g. JAX, we can easily compute such higher-order
+# derivatives:
```
```suggestion
# gain?
# 
# Nothing, actually, for the full reconstruction! Quite the opposite, we spent :math:`2R`
# evaluations of :math:`E` on each part, that is :math:`4R` evaluations overall to obtain a
# description of the full function :math:`E`! This is way more than the :math:`2R+1` 
# evaluations needed for the full reconstructions from the beginning.
#
# However, remember that we set out to compute derivatives of :math:`E` at :math:`0`, so that
# for derivatives of odd/even order only the odd/even reconstruction is required.
# Using an autodifferentiation framework, e.g. JAX, we can easily compute such higher-order
# derivatives:
```

Optional! But I find line breaks help emphasize the message (note that online texts, such as online news articles, tend to have a _lot_ more line breaks than printed media!)

![image](https://user-images.githubusercontent.com/2959003/130422524-ef9d3faa-91f4-4a92-9cfc-cf9996e0b9bd.png)

dwierichs replied:
I agree, it's often much nicer to read. Still trying to do this more automatically :)

josh146 reviewed (changes requested):
Agree with @mariaschuld --- the revised demo structure is a huge improvement! The story is clear, there is a linear progression, and the reader comes away feeling a sense of achievement.

In particular, I just want to say I love the duality of the last section - using JAX _and_ implementing the equation explicitly.

Great work, from my point of view this review has transitioned to minor polishing :)

dwierichs commented:
I implemented basically all suggestions right away. 
For the comparison of the two differentiation methods, I added some more detail on how they work, and what their respective advantage is. A shortened version of this is also in the intro, already mentioning the trade-off between autodiff cost at higher orders on one hand and manual diff work on the other hand.
I hope the legend for the summed reconstruction from the even and odd part helps.

re: binary files is WIP

dwierichs commented:
Thanks a lot for the repeated reviews, those were really helpful! :)
Fixed the binary files and would be ready for re-review of the last changes / merging.

mariaschuld commented on demonstrations/tutorial_general_parshift.py:
```diff
@@ -0,0 +1,880 @@
+#
+# Here, :math:`a_{\ell}` and :math:`b_{\ell}` are the *Fourier coefficients*.
+# If you would like to understand this a bit better still, have a read of
+# `:mod:`~.pennylane.fourier` and remember to check out the
```
This is not rendering correctly, but I also don't know why :)

dwierichs replied:
Ah, the initial ` :)

mariaschuld commented on demonstrations/tutorial_general_parshift.py:
```diff
@@ -0,0 +1,880 @@
+# with the order and we always reconstruct the *same* type of function, namely Fourier series,
+# so that computing the respective derivatives once manually and coding up the resulting 
+# coefficients of the parameter-shift rule pays off in the long run. This is the strength of the
+# second approach.
```
Very nice explanation!

mariaschuld reviewed (approved):
Nice demo, and the changes were well-implemented. I only had a superficial look, and saw one rendering issue. Once that is fixed, happy for this to be merged!

dwierichs commented:
I found another code output from the new legend, fixed the rendering issue and updated the _posted_-date. Ready on my end :)

josh146 commented on demonstrations/tutorial_general_parshift.py:
```diff
@@ -0,0 +1,880 @@
+
+Cost functions arising from quantum gates
+-----------------------------------------
+We start our investigations with a cost function that arises from measuring the expectation
```
```suggestion
We start our investigation by considering a cost function that arises from measuring the expectation
```

josh146 commented on demonstrations/tutorial_general_parshift.py:
```diff
@@ -0,0 +1,880 @@
+Cost functions arising from quantum gates
+-----------------------------------------
+We start our investigations with a cost function that arises from measuring the expectation
+value of an observable in a quantum state created with a parametrized quantum operation
+that depends on a single variational parameter :math:`x`.
```
```suggestion
value of an observable on a quantum state, created with a parametrized quantum operation
that depends on a single variational parameter :math:`x`.
```
I'm not 100% sure here, but is it on or in?

dwierichs replied:
I'd favor "in".

josh146 commented on demonstrations/tutorial_general_parshift.py:
```diff
@@ -0,0 +1,880 @@
+
+
+###############################################################################
+# The second-order derivative takes a similar form but we have to take care of the evaluation at
```
```suggestion
# The second-order derivative takes a similar form, but we have to take care of the evaluation at
```

josh146 reviewed (approved):
🎉 💯
